### PR TITLE
[MLIR] Improve the compiler driver failure reporting

### PR DESCRIPTION
**Pull request by grwlf**
In this PR we add the following functionality to the CompilerDriver:
* [x] Print the name of failed pass along with the name of the pipeline
* [x] Dump the failing IR into the workspace with the "_FAILED" suffix

### CODE REVIEW

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 1]
```diff
@@ -364,7 +364,12 @@ LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, Modu
     auto afterPassFailedCallback = [&](Pass *pass, Operation *op) {
         auto res = passPipelineNames.find(pass);
         assert(res != passPipelineNames.end() && "Unexpected pass");
-        options.diagnosticStream << "While processing pipeline: " << res->second << "\n";
+        options.diagnosticStream << "While processing '" << pass->getName() << "' pass "
+                                 << "of the '" << res->second << "' pipeline\n";
+        llvm::raw_string_ostream s{outputs[res->second]};
+        s << *op;
+        dumpToFile(options, output.nextPipelineDumpFilename(res->second + "_FAILED"),
+                   outputs[res->second]);
```
Comment (dime10) [inline mlir/lib/Driver/CompilerDriver.cpp:372]: Should this be guarded by a verbose or debug flag (or something similar)?

**Review by dime10 — APPROVED** [review]
Love it!

**Comment by dime10** [comment]
Changelog :)

File: mlir/lib/Driver/CompilerDriver.cpp (hunk 1 above)
Comment (grwlf) [reply mlir/lib/Driver/CompilerDriver.cpp]: Indeed. I wrapped it with keep_intermediate, thanks!

**Pull request merged**